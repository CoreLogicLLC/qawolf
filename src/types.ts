export type Action =
  | 'click'
  | 'fill'
  | 'goBack'
  | 'goForward'
  | 'goto'
  | 'press'
  | 'reload'
  | 'scroll'
  | 'selectOption'
  | 'type';

export type BrowserName = 'chromium' | 'firefox' | 'webkit';

export type Callback<S = void, T = void> = (data?: S) => T;

export interface Doc {
  attrs: Record<string, string>;
  name: string;
}

export interface PageEvent {
  frameIndex?: number;
  frameSelector?: string;
  name: string;
  page: number;
  time: number;
  value?: string | ScrollValue | null;
}

export interface WindowEvent extends PageEvent {
  name: WindowEventName;
}

export type WindowEventName =
  | 'goBack'
  | 'goForward'
  | 'goto'
  | 'reload';

export interface ElementEvent extends PageEvent {
  isTrusted: boolean;
  name: ElementEventName;
  selector: string;
  target: Doc;
}

export type ElementEventName =
  | 'change'
  | 'click'
  | 'input'
  | 'keydown'
  | 'keyup'
  | 'mousedown'
  | 'paste'
  | 'scroll';

export interface InputEvent extends ElementEvent {
  name: 'input';
  value: string | null;
}

export interface KeyEvent extends ElementEvent {
  name: 'keydown' | 'keyup';
  // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code
  value: string;
}

export interface LogEvent {
  level: string;
  message: string;
}

export interface PasteEvent extends ElementEvent {
  name: 'paste';
  value: string;
}

export interface ScrollEvent extends ElementEvent {
  name: 'scroll';
  value: ScrollValue;
}

export type ScrollValue = {
  x: number;
  y: number;
};

export interface Step {
  action: Action;
  event: PageEvent;
  // needed to build selector key
  index: number;
  value?: StepValue;
}

export type StepValue = string | ScrollValue | null | undefined;

export interface Workflow {
  device?: string;
  name: string;
  steps: Step[];
  url: string;
}
