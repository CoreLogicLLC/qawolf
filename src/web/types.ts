export type SelectorPart = {
  name: 'css' | 'text';
  body: string;
};

export type ParsedSelector = {
  capture?: boolean;
  parts: SelectorPart[];
};

export type Evaluator = {
  createTextSelector(element: Element): string | undefined;

  isVisible(element: Element): boolean;

  querySelector(selector: ParsedSelector, root: Node): HTMLElement;
};
