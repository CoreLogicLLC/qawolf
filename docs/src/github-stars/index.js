module.exports = function(context) {
  const { siteConfig } = context;
  const { themeConfig } = siteConfig;
  const { github } = themeConfig || {};

  if (!github || !github.repo) {
    throw new Error(`You need to specify 'github.repo' in 'themeConfig'`);
  }

  return {
    name: "github-stars",

    injectHtmlTags() {
      return {
        headTags: [
          {
            tagName: "script",
            innerHTML: `
            function htmlToElement(html) {
              var template = document.createElement('template');
              html = html.trim(); // Never return a text node of whitespace as the result
              template.innerHTML = html;
              return template.content.firstChild;
            }

            var starsIntervalId = setInterval(() => {
              const stars = document.querySelector(".github-wrapper");
              if (stars) return;

              const navBar = document.querySelector(".navbar__items.navbar__items--right");
              if (!navBar) return;

              var starsLink = htmlToElement('<div class="github-wrapper"><a class="github-button" href="https://github.com/${github.repo}" data-icon="octicon-star" data-size="large" data-show-count="true" aria-label="Star ${github.repo} on GitHub">Star</a></div>');
              navBar.append(starsLink);
              
              var starsScript = document.createElement('script');
              starsScript.src = "https://buttons.github.io/buttons.js";
              document.head.append(starsScript);
            }, 100);`
          }
        ]
      };
    }
  };
};
