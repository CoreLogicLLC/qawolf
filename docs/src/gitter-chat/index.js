module.exports = function(context) {
  const { siteConfig } = context;
  const { themeConfig } = siteConfig;
  const { gitter } = themeConfig || {};

  if (!gitter || !gitter.room) {
    throw new Error(`You need to specify 'gitter.room' in 'themeConfig'`);
  }

  return {
    name: "gitter-chat",

    injectHtmlTags() {
      return {
        headTags: [
          {
            tagName: "script",
            innerHTML: `
            ((window.gitter = {}).chat = {}).options = {
              room: '${gitter.room}'
            };
            `
          },
          {
            tagName: "script",
            attributes: {
              async: true,
              defer: true,
              src: "https://sidecar.gitter.im/dist/sidecar.v1.js"
            }
          }
        ]
      };
    }
  };
};
