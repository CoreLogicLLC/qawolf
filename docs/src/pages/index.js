function Home() {
  if (typeof window !== "undefined") {
    // redirect to Home page
    window.location.href = "https://www.qawolf.com";
  }

  return null;
}

export default Home;
