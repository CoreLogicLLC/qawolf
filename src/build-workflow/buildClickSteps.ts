import { isInputTarget } from './target';
import { ElementEvent, Step } from '../types';

const CLICK_EVENTS = ['click', 'mousedown'];

const shouldIncludeClickEvent = (
  event: ElementEvent,
  previousEvent?: ElementEvent,
): boolean => {
  // ignore system initiated clicks
  if (!event.isTrusted) return false;

  // ignore non-click events
  if (!CLICK_EVENTS.includes(event.name)) return false;

  // ignore clicks on selects
  if (event.target.name === 'select') return false;

  // skip system-initiated clicks triggered by a key press
  // ex. "Enter" triggers a click on a submit input
  if (
    previousEvent &&
    ['change', 'keydown', 'keyup'].includes(previousEvent.name) &&
    event.time - previousEvent.time < 50
  ) {
    return false;
  }

  return true;
};

const groupClickEvents = (events: ElementEvent[]): ElementEvent[][] => {
  const clickGroups = [];

  let group: ElementEvent[] = [];

  events.forEach((event, i) => {
    const previousEvent = events[i - 1];

    if (!shouldIncludeClickEvent(event, previousEvent)) return;

    // always group clicks with the previous mousedown/click events
    // a click will follow a mousedown if the mouse was released within the same element
    // a click will follow a click if it propagated to a higher element (a click on a label propagates to the input)
    if (!group.length || event.name === 'click') {
      group.push(event);
      return;
    }

    // start a new group
    clickGroups.push(group);
    group = [event];
  });

  if (group.length) {
    // append last group
    clickGroups.push(group);
  }

  return clickGroups;
};

export const buildClickSteps = (events: ElementEvent[]): Step[] => {
  const groupedClickEvents = groupClickEvents(events);
  const steps: Step[] = [];

  groupedClickEvents.forEach((events) => {
    let event = events[0] as ElementEvent;

    const inputEvent = events.find((event) => isInputTarget(event.target));
    if (inputEvent) {
      // if an event in the group is on an input, assume the click propagated
      // to an element like a checkbox or radio, which is most accurate target
      event = inputEvent;
    }

    steps.push({
      action: 'click',
      event,
      index: steps.length,
    });
  });

  return steps;
};
