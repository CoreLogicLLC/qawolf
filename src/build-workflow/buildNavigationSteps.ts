import { Action, WindowEvent, Step } from '../types';

export const buildNavigationSteps = (events: WindowEvent[]): Step[] => {
  return events.map((event, index) => ({
    action: (event.name as Action),
    event,
    index,
    value: event.value,
  }));
};
