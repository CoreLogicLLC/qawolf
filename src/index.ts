#!/usr/bin/env node

import * as qawolf from './qawolf';
import { Registry } from './utils';

// support: const qawolf = require("qawolf");
// support: import { ... } from "qawolf";
export * from './qawolf';

// support: import qawolf from "qawolf"
export default qawolf;

// must do this here to prevent circular dependency
Registry.instance().setQawolf(qawolf);

if (!module.parent) {
  // run the cli when this is the root module
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { runCli } = require('./cli/cli');
  runCli();
}
