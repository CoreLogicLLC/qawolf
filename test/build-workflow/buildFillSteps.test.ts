import { loadFixtures } from '../loadFixtures';
import { buildFillSteps } from '../../src/build-workflow';
import { ElementEvent } from '../../src/types';

describe('buildFillSteps', () => {
  it('clicking and typing in a text input creates a single fill step', async () => {
    const fillTextInput = await loadFixtures('fillTextInput');

    const steps = buildFillSteps(fillTextInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and editing existing text in a text input creates a single fill step', async () => {
    const fillExistingTextInput = await loadFixtures('fillExistingTextInput');

    const steps = buildFillSteps(fillExistingTextInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a textarea creates a single fill step', async () => {
    const fillTextarea = await loadFixtures('fillTextarea');

    const steps = buildFillSteps(fillTextarea.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a contenteditable creates a single fill step', async () => {
    const fillContentEditable = await loadFixtures('fillContentEditable');

    const steps = buildFillSteps(fillContentEditable.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and editing existing text in a contenteditable creates a single fill step', async () => {
    const fillExistingContentEditable = await loadFixtures('fillExistingContentEditable');

    const steps = buildFillSteps(fillExistingContentEditable.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a search input creates a single fill step', async () => {
    const fillSearchInput = await loadFixtures('fillSearchInput');

    const steps = buildFillSteps(fillSearchInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser date input creates a single fill step', async () => {
    const fillDateInput = await loadFixtures('fillDateInput');

    const steps = buildFillSteps(fillDateInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser datetime-local input creates a single fill step', async () => {
    const fillDateTimeLocalInput = await loadFixtures('fillDateTimeLocalInput');

    const steps = buildFillSteps(fillDateTimeLocalInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser time input creates a single fill step', async () => {
    const fillTimeInput = await loadFixtures('fillTimeInput');

    const steps = buildFillSteps(fillTimeInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser week input creates a single fill step', async () => {
    const fillWeekInput = await loadFixtures('fillWeekInput');

    const steps = buildFillSteps(fillWeekInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser month input creates a single fill step', async () => {
    const fillMonthInput = await loadFixtures('fillMonthInput');

    const steps = buildFillSteps(fillMonthInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser number input creates a single fill step', async () => {
    const fillNumberInput = await loadFixtures('fillNumberInput');

    const steps = buildFillSteps(fillNumberInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser email input creates a single fill step', async () => {
    const fillEmailInput = await loadFixtures('fillEmailInput');

    const steps = buildFillSteps(fillEmailInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser password input creates a single fill step', async () => {
    const fillPasswordInput = await loadFixtures('fillPasswordInput');

    const steps = buildFillSteps(fillPasswordInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser phone number input creates a single fill step', async () => {
    const fillPhoneInput = await loadFixtures('fillPhoneInput');

    const steps = buildFillSteps(fillPhoneInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking and typing in a browser url input creates a single fill step', async () => {
    const fillUrlInput = await loadFixtures('fillUrlInput');

    const steps = buildFillSteps(fillUrlInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('picking a color from a browser color input creates a single fill step', async () => {
    const pickColorInput = await loadFixtures('pickColorInput');

    const steps = buildFillSteps(pickColorInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('picking a number on a browser range input creates a single fill step', async () => {
    const pickRangeInput = await loadFixtures('pickRangeInput');

    const steps = buildFillSteps(pickRangeInput.events as ElementEvent[]);

    expect(steps.length).toBe(1);
    expect(steps[0]).toMatchSnapshot();
  });

  it('clicking a checkbox does not create a fill step', async () => {
    const selectCheckboxInput = await loadFixtures('selectCheckboxInput');

    const steps = buildFillSteps(selectCheckboxInput.events as ElementEvent[]);

    expect(steps.length).toBe(0);
  });

  it('clicking a radio button does not create a fill step ', async () => {
    const selectRadioInput = await loadFixtures('selectRadioInput');

    const steps = buildFillSteps(selectRadioInput.events as ElementEvent[]);

    expect(steps.length).toBe(0);
  });
});
