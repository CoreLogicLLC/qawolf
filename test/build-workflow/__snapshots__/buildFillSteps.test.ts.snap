// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildFillSteps clicking and editing existing text in a contenteditable creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"content-editable\\"]",
    "target": Object {
      "attrs": Object {
        "contenteditable": "true",
        "data-qa": "content-editable",
      },
      "children": Array [
        Object {
          "content": "modified text",
          "type": "text",
        },
        Object {
          "attrs": Object {},
          "children": Array [
            Object {
              "attrs": Object {},
              "children": Array [],
              "name": "br",
              "type": "tag",
              "voidElement": true,
            },
          ],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {},
          "children": Array [
            Object {
              "content": "multiple lines",
              "type": "text",
            },
          ],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "name": "div",
      "type": "tag",
      "voidElement": false,
    },
    "time": 1594313076075,
    "value": "modified text


multiple lines",
  },
  "index": 0,
  "value": "modified text


multiple lines",
}
`;

exports[`buildFillSteps clicking and editing existing text in a text input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-text-input-filled\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-text-input-filled",
        "placeholder": "Filled text input",
        "type": "text",
        "value": "initial text",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594311846976,
    "value": "changed text",
  },
  "index": 0,
  "value": "changed text",
}
`;

exports[`buildFillSteps clicking and typing in a browser date input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-date-picker\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-date-picker",
        "type": "date",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594312367406,
    "value": "1999-04-10",
  },
  "index": 0,
  "value": "1999-04-10",
}
`;

exports[`buildFillSteps clicking and typing in a browser datetime-local input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-datetime-local-picker\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-datetime-local-picker",
        "type": "datetime-local",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594312714472,
    "value": "2005-02-20T01:50",
  },
  "index": 0,
  "value": "2005-02-20T01:50",
}
`;

exports[`buildFillSteps clicking and typing in a browser email input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-email-input\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-email-input",
        "placeholder": "Email input",
        "type": "email",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594312113762,
    "value": "me@example.net",
  },
  "index": 0,
  "value": "me@example.net",
}
`;

exports[`buildFillSteps clicking and typing in a browser month input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-month-picker\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-month-picker",
        "type": "month",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594312547159,
    "value": "2020-06",
  },
  "index": 0,
  "value": "2020-06",
}
`;

exports[`buildFillSteps clicking and typing in a browser number input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-number-input\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-number-input",
        "placeholder": "Number input",
        "type": "number",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594311943233,
    "value": "123",
  },
  "index": 0,
  "value": "123",
}
`;

exports[`buildFillSteps clicking and typing in a browser password input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-password-input\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-password-input",
        "placeholder": "Password input",
        "type": "password",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594311900120,
    "value": "password",
  },
  "index": 0,
  "value": "password",
}
`;

exports[`buildFillSteps clicking and typing in a browser phone number input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-tel-input\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-tel-input",
        "placeholder": "Telephone input",
        "type": "tel",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594312150820,
    "value": "5551264585",
  },
  "index": 0,
  "value": "5551264585",
}
`;

exports[`buildFillSteps clicking and typing in a browser time input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-time-picker\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-time-picker",
        "type": "time",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594312672013,
    "value": "16:19",
  },
  "index": 0,
  "value": "16:19",
}
`;

exports[`buildFillSteps clicking and typing in a browser url input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-url-input\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-url-input",
        "placeholder": "URL input",
        "type": "url",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594312203806,
    "value": "https://qawolf.com",
  },
  "index": 0,
  "value": "https://qawolf.com",
}
`;

exports[`buildFillSteps clicking and typing in a browser week input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-week-picker\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-week-picker",
        "type": "week",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594312510328,
    "value": "1997-W09",
  },
  "index": 0,
  "value": "1997-W09",
}
`;

exports[`buildFillSteps clicking and typing in a contenteditable creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"draftjs\\"] [contenteditable=\\"true\\"]",
    "target": Object {
      "attrs": Object {
        "aria-describedby": "placeholder-da09n",
        "class": "notranslate public-DraftEditor-content",
        "contenteditable": "true",
        "role": "textbox",
        "spellcheck": "false",
        "style": "outline: none; user-select: text; white-space: pre-wrap; overflow-wrap: break-word;",
      },
      "children": Array [
        Object {
          "attrs": Object {
            "data-contents": "true",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "class": "",
                "data-block": "true",
                "data-editor": "da09n",
                "data-offset-key": "aj4j3-0-0",
              },
              "children": Array [
                Object {
                  "attrs": Object {
                    "class": "public-DraftStyleDefault-block public-DraftStyleDefault-ltr",
                    "data-offset-key": "aj4j3-0-0",
                  },
                  "children": Array [
                    Object {
                      "attrs": Object {
                        "data-offset-key": "aj4j3-0-0",
                      },
                      "children": Array [
                        Object {
                          "attrs": Object {
                            "data-text": "true",
                          },
                          "children": Array [
                            Object {
                              "content": "QAWolf is good",
                              "type": "text",
                            },
                          ],
                          "name": "span",
                          "type": "tag",
                          "voidElement": false,
                        },
                      ],
                      "name": "span",
                      "type": "tag",
                      "voidElement": false,
                    },
                  ],
                  "name": "div",
                  "type": "tag",
                  "voidElement": false,
                },
              ],
              "name": "div",
              "type": "tag",
              "voidElement": false,
            },
            Object {
              "attrs": Object {
                "class": "",
                "data-block": "true",
                "data-editor": "da09n",
                "data-offset-key": "dnr0o-0-0",
              },
              "children": Array [
                Object {
                  "attrs": Object {
                    "class": "public-DraftStyleDefault-block public-DraftStyleDefault-ltr",
                    "data-offset-key": "dnr0o-0-0",
                  },
                  "children": Array [
                    Object {
                      "attrs": Object {
                        "data-offset-key": "dnr0o-0-0",
                      },
                      "children": Array [
                        Object {
                          "attrs": Object {
                            "data-text": "true",
                          },
                          "children": Array [
                            Object {
                              "content": "Not bad",
                              "type": "text",
                            },
                          ],
                          "name": "span",
                          "type": "tag",
                          "voidElement": false,
                        },
                      ],
                      "name": "span",
                      "type": "tag",
                      "voidElement": false,
                    },
                  ],
                  "name": "div",
                  "type": "tag",
                  "voidElement": false,
                },
              ],
              "name": "div",
              "type": "tag",
              "voidElement": false,
            },
          ],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "name": "div",
      "type": "tag",
      "voidElement": false,
    },
    "time": 1594313142633,
    "value": "QAWolf is good
Not bad",
  },
  "index": 0,
  "value": "QAWolf is good
Not bad",
}
`;

exports[`buildFillSteps clicking and typing in a search input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-search-input\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-search-input",
        "placeholder": "Search input",
        "type": "search",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594311791266,
    "value": "search text",
  },
  "index": 0,
  "value": "search text",
}
`;

exports[`buildFillSteps clicking and typing in a text input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-text-input\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-text-input",
        "placeholder": "Text input",
        "type": "text",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594311579128,
    "value": "Lazy dog",
  },
  "index": 0,
  "value": "Lazy dog",
}
`;

exports[`buildFillSteps clicking and typing in a textarea creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-textarea\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-textarea",
        "placeholder": "Textarea",
      },
      "children": Array [],
      "name": "textarea",
      "type": "tag",
      "voidElement": false,
    },
    "time": 1594312006659,
    "value": "Multi

Line

text!",
  },
  "index": 0,
  "value": "Multi

Line

text!",
}
`;

exports[`buildFillSteps picking a color from a browser color input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-color-picker\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-color-picker",
        "type": "color",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594313468216,
    "value": "#ec1313",
  },
  "index": 0,
  "value": "#ec1313",
}
`;

exports[`buildFillSteps picking a number on a browser range input creates a single fill step 1`] = `
Object {
  "action": "fill",
  "event": Object {
    "isTrusted": true,
    "name": "input",
    "page": 0,
    "selector": "[data-qa=\\"html-range-picker\\"]",
    "target": Object {
      "attrs": Object {
        "data-qa": "html-range-picker",
        "max": "10",
        "min": "0",
        "type": "range",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "time": 1594313495655,
    "value": "4",
  },
  "index": 0,
  "value": "4",
}
`;
